src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -1075,17 +1077,20 @@ private BoundExpression CreateAndPopulateList(BoundCollectionExpression node, Ty
 
                             // If collection has a struct enumerator but doesn't implement ICollection<T>
                             // then manual `foreach` is always more efficient then using `AddRange` method
-                            if (enumeratorInfo.GetEnumeratorInfo.Method.ReturnType.IsValueType &&
-                                !enumeratorInfo.CollectionType.ImplementsInterface(iCollectionOfElementType, ref discardedUseSiteInfo))
+                            if (enumeratorInfo.GetEnumeratorInfo.Method.ReturnType.IsValueType)
                             {
-                                return false;
+                                conversion = _compilation.Conversions.ClassifyBuiltInConversion(enumeratorInfo.CollectionType, iCollectionOfElementType, isChecked: false, ref discardedUseSiteInfo);
+                                if (!(conversion.Kind is ConversionKind.Identity or ConversionKind.ImplicitReference))
```

Comment:
nit: since the below conversion.Kind check for the AddRangeMethod parameter type is very similar in purpose, I would prefer if they were written in the same way, revising one or both as desired.